File: src/kernel/archs/x86_64/gdt.c
#include <errno.h>
#include <hal>
#include <kmalloc>
#include <logger>
#include <string.h>
#include <sync>

#include "ctx.h"
#include "gdt.h"

static Gdt gdt = {0};
static Tss tss = {0};

static Spinlock lock = SPINLOCK_INIT;

static GdtDescriptor gdt_desc = {
    .size = sizeof(gdt) - 1,
    .offset = (uintptr_t)&gdt,
};

static void gdt_lazy_init(GdtEntry self[static 1], uint8_t access, uint8_t flags)
{
    memset(self, 0, sizeof(GdtEntry));

    if (access == 0 && flags == 0)
    {
        return;
    }

    self->access = access | GDT_ACCESS_PRESENT | GDT_ACCESS_READ_WRITE | GDT_ACCESS_DESCRIPTOR;
    self->flags = flags | GDT_FLAGS_GRANULARITY;

    self->base_high = 0;
    self->base_middle = 0;
    self->base_low = 0;

    self->limit_low = 0xffff;
    self->limit_high = 0x0f;
}

void gdt_load_tss(Tss *tss)
{
    uintptr_t tss_ptr = (uintptr_t)tss;

    spinlock_acquire(&lock);

    gdt.tss_entry = (TssEntry){
        .length = sizeof(TssEntry),

        .base_low = tss_ptr & 0xffff,
        .base_mid = (tss_ptr >> 16) & 0xff,
        .base_high = (tss_ptr >> 24) & 0xff,
        .base_upper = tss_ptr >> 32,

        .flags1 = TSS_FLAGS_PRESENT | TSS_FLAGS_64BITS_AVAILABLE,
        .flags2 = 0,

        .reserved = 0,
    };

    spinlock_release(&lock);
}

uintptr_t gdt_descriptor(void)
{
    return (uintptr_t)&gdt_desc;
}

void gdt_init(void)
{
    gdt_lazy_init(&gdt.entries[GDT_NULL_DESC], 0, 0);

    gdt_lazy_init(&gdt.entries[GDT_KERNEL_CODE], GDT_ACCESS_EXE, GDT_FLAGS_LONG_MODE);
    gdt_lazy_init(&gdt.entries[GDT_KERNEL_DATA], 0, GDT_FLAGS_SIZE);

    gdt_lazy_init(&gdt.entries[GDT_USER_DATA], GDT_ACCESS_USER, GDT_FLAGS_SIZE);
    gdt_lazy_init(&gdt.entries[GDT_USER_CODE], GDT_ACCESS_USER | GDT_ACCESS_EXE, GDT_FLAGS_LONG_MODE);
    gdt_load_tss(NULL);

    gdt_flush(gdt_descriptor());
    log$("GDT loaded");

    tss_flush();
    log$("TSS loaded");
}

void gdt_init_tss(void)
{
    Alloc heap = kmalloc_acquire();

    void *ist1 = heap.malloc(KERNEL_STACK_SIZE);
    void *ist0 = heap.malloc(KERNEL_STACK_SIZE);
    void *rsp0 = heap.malloc(KERNEL_STACK_SIZE);

    if (IS_ERR(ist1) || IS_ERR(ist0) || IS_ERR(rsp0))
    {
        error$("Failed to allocate memory for TSS stacks");
        hal_panic();
    }

    tss.ist[1] = (uintptr_t)(ist1 + KERNEL_STACK_SIZE);
    tss.ist[0] = (uintptr_t)(ist0 + KERNEL_STACK_SIZE);
    tss.rsp[0] = (uintptr_t)(rsp0 + KERNEL_STACK_SIZE);

    log$("TSS initialized");
    gdt_load_tss(&tss);
    tss_flush();
}

File: src/kernel/archs/x86_64/gdt.zig
const std = @import("std");
const log = std.log.scoped(.gdt);
const PageAllocator = @import("./pmm.zig").PageAllocator;
const kib = @import("utils").kib;

extern fn gdt_flush(addr: u64) void;
extern fn tss_flush() void;

pub const GdtType = enum(u64) {
    Null = 0,
    KernelCode = 1,
    KernelData = 2,
    UserData = 3,
    UserCode = 4,

    Len = 5,
};

const GdtAccess = enum(u8) {
    ReadWrite = (1 << 1),
    Executable = (1 << 3),
    Descriptor = (1 << 4),
    User = (3 << 5),
    Present = (1 << 7),
};

const GdtFlags = enum(u8) {
    LongMode = (1 << 1),
    Size = (1 << 2),
    Granularity = (1 << 3),
};

const TssFlags = enum(u8) {
    Present = (1 << 7),
    LongModeAvailable = (0x9),
};

const GdtEntry = packed struct {
    limit_low: u16,
    base_low: u16,
    base_middle: u8,
    access: u8,
    limit_high: u4,
    flags: u4,
    base_high: u8,

    pub fn fast_init(access: u8, flags: u8) GdtEntry {
        var entry = std.mem.zeroes(GdtEntry);

        entry.access = access | @intFromEnum(GdtAccess.Present) | @intFromEnum(GdtAccess.ReadWrite) | @intFromEnum(GdtAccess.Descriptor);
        entry.flags = @intCast(flags | @intFromEnum(GdtFlags.Granularity));

        entry.limit_low = 0xffff;
        entry.limit_high = 0x0f;

        return entry;
    }
};

const Tss = packed struct {
    reserved0: u32,
    rsp: packed struct {
        rsp0: u64,
        rsp1: u64,
        rsp2: u64,
    },
    reserved1: u64,
    ist: packed struct {
        ist0: u64,
        ist1: u64,
        ist2: u64,
        ist3: u64,
        ist4: u64,
        ist5: u64,
        ist6: u64,
    },
    reserved2: u80,
    iopb_offset: u16,
};

const TssEntry = packed struct {
    length: u16,
    base_low: u16,
    base_middle: u8,
    flags: u16,
    base_high: u8,
    base_upper: u32,
    reserved: u32 = 0,

    pub fn from_addr(addr: u64) TssEntry {
        return .{
            .length = @intCast(@sizeOf(Tss)),
            .base_low = @intCast(addr & 0xffff),
            .base_middle = @intCast((addr >> 16) & 0xff),
            .base_high = @intCast((addr >> 24) & 0xff),
            .base_upper = @intCast(addr >> 32),
            .flags = @intCast(@intFromEnum(TssFlags.Present) | @intFromEnum(TssFlags.LongModeAvailable)),
        };
    }
};

const Gdt = extern struct {
    entries: [@intFromEnum(GdtType.Len)]GdtEntry align(1),
    tss: TssEntry align(1),
};

const GdtDescriptor = packed struct {
    limit: u16,
    base: u64,

    pub fn load(_gdt: *Gdt) GdtDescriptor {
        return .{
            .limit = @intCast(@sizeOf(Gdt) - 1),
            .base = @intFromPtr(_gdt),
        };
    }

    pub fn apply(self: *const GdtDescriptor) void {
        gdt_flush(@intFromPtr(self));
    }
};

var gdt: Gdt = std.mem.zeroes(Gdt);
var tss: Tss = std.mem.zeroes(Tss);

pub fn setup() void {
    gdt.entries[@intFromEnum(GdtType.Null)] = std.mem.zeroes(GdtEntry);

    gdt.entries[@intFromEnum(GdtType.KernelCode)] = GdtEntry.fast_init(@intFromEnum(GdtAccess.Executable), @intFromEnum(GdtFlags.LongMode));
    gdt.entries[@intFromEnum(GdtType.KernelData)] = GdtEntry.fast_init(0, @intFromEnum(GdtFlags.LongMode));
    gdt.entries[@intFromEnum(GdtType.UserData)] = GdtEntry.fast_init(@intFromEnum(GdtAccess.User), @intFromEnum(GdtFlags.LongMode));
    gdt.entries[@intFromEnum(GdtType.UserCode)] = GdtEntry.fast_init(@intFromEnum(GdtAccess.Executable) | @intFromEnum(GdtAccess.User), @intFromEnum(GdtFlags.LongMode));

    gdt.tss = TssEntry.from_addr(0);

    GdtDescriptor.load(&gdt).apply();

    tss_flush();
    log.debug("Gdt loaded", .{});
}

pub fn setup_tss() !void {
    var allocator = PageAllocator.new();
    const alloc = allocator.allocator();

    const ist0 = (try alloc.alloc(u8, kib(1))).ptr;
    const ist1 = (try alloc.alloc(u8, kib(1))).ptr;
    const rsp0 = (try alloc.alloc(u8, kib(1))).ptr;

    tss.ist.ist0 = @intFromPtr(ist0) + kib(1);
    tss.ist.ist1 = @intFromPtr(ist1) + kib(1);
    tss.rsp.rsp0 = @intFromPtr(rsp0) + kib(1);

    gdt.tss = TssEntry.from_addr(@intFromPtr(&tss));
    tss_flush();
}
